sherpa/csrc/byte_util.cc

```diff
@@ -0,0 +1,212 @@
+  max_token_ = 422;  // the max number in above table
+  token2byte_ =
+      std::vector<int16_t>(max_token_ + 1, -1);  // the max token in byte2token_
+                                                 // is 422, so we set the length
+                                                 // of token2bytes_ 423.
+  for (size_t i = 0; i < byte2token_.size(); ++i) {
+    token2byte_[byte2token_[i]] = i;
+  }
+}
+
+std::string ByteUtil::Encode(const std::string &str) const {
+  std::ostringstream oss;
+  const uint8_t *p = reinterpret_cast<const uint8_t *>(str.data());
+  for (size_t i = 0; i < str.size(); ++i) {
+    oss << CodePointToUTF8String(byte2token_[p[i]]);
+  }
+  return oss.str();
+}
+
+std::string ByteUtil::Decode(const std::string &str) const {
+  std::vector<uint8_t> bytes;
+  UTF8StringToBytes(str, &bytes);
+  std::vector<int32_t> codes;
+  BytesToCodePoints(bytes.data(), bytes.size(), &codes);
+  std::ostringstream oss;
+  for (size_t i = 0; i < codes.size(); ++i) {
+    oss << CodePointToUTF8String(codes[i]);
+  }
+  return oss.str();
+}
+
+void ByteUtil::UTF8StringToBytes(const std::string &str,
+                                 std::vector<uint8_t> *bytes) const {
+  const char *data = str.data();
+  const size_t length = str.size();
+  for (size_t i = 0; i < length; /* no update */) {
+    int32_t c = data[i++] & 0xff;
+    if ((c & 0x80) == 0) {
+      if (c > max_token_ || token2byte_[c] == -1) {
+        SHERPA_LOG(WARNING) << "Skip OOV token, code point : " << c
+                            << " utf8 char : " << CodePointToUTF8String(c);
+        continue;
+      }
+      bytes->push_back(token2byte_[c]);
+    } else {
+      if ((c & 0xc0) == 0x80) {
+        SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+      }
+      int32_t count =
+          (c >= 0xc0) + (c >= 0xe0) + (c >= 0xf0) + (c >= 0xf8) + (c >= 0xfc);
+      int32_t code = c & ((1 << (6 - count)) - 1);
+      while (count != 0) {
+        if (i == length) {
+          SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+        }
+        char cb = data[i++];
+        if ((cb & 0xc0) != 0x80) {
+          SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+        }
+        code = (code << 6) | (cb & 0x3f);
+        count--;
+      }
+      if (code < 0) {
+        // This should not be able to happen.
+        SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+      }
+      if (code > max_token_ || token2byte_[code] == -1) {
+        SHERPA_LOG(WARNING) << "Skip OOV token, code point : " << code
+                            << " utf8 char : " << CodePointToUTF8String(code);
+        continue;
+      }
+      bytes->push_back(token2byte_[code]);
+    }
+  }
+}
+
+void ByteUtil::BytesToCodePoints(uint8_t *bytes, int32_t length,
+                                 std::vector<int32_t> *codes) const {
+  if (length <= 0) return;
+  char *data = reinterpret_cast<char *>(bytes);
+  int32_t idx = 1;  // means start from next byte
+  for (int32_t i = 0; i < length; /* no update */) {
```

Comment:
```suggestion
  codes->clear();
  for (int32_t i = 0; i < length; /* no update */) {
```

Reply:
Can not clear here, invoked recursively.